File: hub/exchange/src/device/attribute.py
from device import DataType
from .parameter import Parameter

class Attribute:
    
    def __init__(self,name, parameters=[], isControllable=True):
        self.name=name
        self.parameters=parameters
        self.isControllable=isControllable

    def __str__(self):
        return "Attribute [name: "+self.name+", parameters: "+str(self.parameters)+"]"

    @property
    def parameters(self):
        return self._parameters
    
    @parameters.setter
    def parameters(self,value):
        if( not isinstance(value,list)):
            raise TypeError('Parameters must be a list of Parameters')
        self._parameters=value

    def addParameter(self,param):
        if(not isinstance(param, Parameter)):
            raise TypeError('{} not a valid Parameter'.format(param))
        self._parameters.append(param) 

File: hub/remote/public/core/device/attribute.js
import $                from 'jquery';
import Component        from '../component';
import DeviceParameter  from './parameter';
import DataType         from './data-type';

import './attribute.css';

class DeviceAttribute extends Component {
    constructor (state, props) {
        super();
        this._state = {
            name        : state.name || 'Unknown Attribute',
            parameters  : state.parameters || []
        };
        this._props = props || { isControllable : false };
    }

    render () {

        this._parameters = this._state.parameters.map((param) => {
            return this._props.isControllable ?
                new DeviceParameter(param.value, param) :
                new DataType(param.value, param);
        });

        this._parameters.forEach((param, i) => {
            param.change((state) => {
                this._state.parameters[i].value = state;
                this.trigger('change', this._state.parameters[i]);
            });
        });

        this._$parameters = $('<div>').addClass('device-params')
            .append(this._parameters.map((param) => {
                return param.render().$el();
            }));


        this._$el
            .empty()
            .addClass('device-attribute')
            .append($('<div>').text(this._state.name).addClass('device-attribute-name'))
            .append(this._$parameters);
        return this;
    }
}

export default DeviceAttribute;

